crypto/serializer/deserializer_pkey.c

```diff
@@ -0,0 +1,327 @@
+
+            *data->object = pkey;
+
+            break;
+        }
+    }
+    /*
+     * We successfully looked through, |*ctx->object| determines if we
+     * actually found something.
+     */
+    return (*data->object != NULL);
+}
+
+static void deser_clean_EVP_PKEY(void *finalize_arg)
+{
+    struct deser_EVP_PKEY_data_st *data = finalize_arg;
+
+    sk_EVP_KEYMGMT_pop_free(data->keymgmts, EVP_KEYMGMT_free);
+    OPENSSL_free(data);
+}
+
+DEFINE_STACK_OF_CSTRING()
+
+struct collected_data_st {
+    struct deser_EVP_PKEY_data_st *process_data;
+    STACK_OF(OPENSSL_CSTRING) *names;
+
+    unsigned int error_occured:1;
+};
+
+static void collect_keymgmt(EVP_KEYMGMT *keymgmt, void *arg)
+{
+    struct collected_data_st *data = arg;
+
+    if (data->error_occured)
+        return;
+
+    data->error_occured = 1;         /* Assume the worst */
+
+    if (!EVP_KEYMGMT_up_ref(keymgmt) /* ref++ */)
+        return;
+    if (sk_EVP_KEYMGMT_push(data->process_data->keymgmts, keymgmt) <= 0) {
+        EVP_KEYMGMT_free(keymgmt); /* ref-- */
+        return;
+    }
+
+    data->error_occured = 0;         /* All is good now */
+}
+
+static void collect_name(const char *name, void *arg)
+{
+    struct collected_data_st *data = arg;
+
+    if (data->error_occured)
+        return;
+
+    data->error_occured = 1;         /* Assume the worst */
+
+    if (sk_OPENSSL_CSTRING_push(data->names, name) <= 0)
+        return;
+
+    data->error_occured = 0;         /* All is good now */
+}
+
+OSSL_DESERIALIZER_CTX *
+OSSL_DESERIALIZER_CTX_new_by_EVP_PKEY(EVP_PKEY **pkey, OPENSSL_CTX *libctx,
+                                      const char *propquery)
+{
+    OSSL_DESERIALIZER_CTX *ctx = NULL;
+    struct collected_data_st *data = NULL;
+    size_t i, end_i;
+
+    if ((ctx = OSSL_DESERIALIZER_CTX_new()) == NULL
+        || (data = OPENSSL_zalloc(sizeof(*data))) == NULL
+        || (data->process_data =
+            OPENSSL_zalloc(sizeof(*data->process_data))) == NULL
+        || (data->process_data->keymgmts
+            = sk_EVP_KEYMGMT_new_null()) == NULL
+        || (data->names = sk_OPENSSL_CSTRING_new_null()) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+    data->process_data->object = (void **)pkey;
+
+    /* First, find all keymgmts to form goals */
+    EVP_KEYMGMT_do_all_provided(libctx, collect_keymgmt, data);
+
+    if (data->error_occured)
+        goto err;
+
+    /*
+     * Then, use the names of those keymgmts to find the first set of
+     * deserializers.
+     */
+    ERR_set_mark();
+    end_i = sk_EVP_KEYMGMT_num(data->process_data->keymgmts);
+    for (i = 0; i < end_i; i++) {
+        EVP_KEYMGMT *keymgmt =
+            sk_EVP_KEYMGMT_value(data->process_data->keymgmts, i);
+        size_t j;
+        OSSL_DESERIALIZER *deser = NULL;
+
+        EVP_KEYMGMT_names_do_all(keymgmt, collect_name, data);
+
+        for (j = sk_OPENSSL_CSTRING_num(data->names); j-- > 0;) {
```

Comment:
Is this the best way to iterate a stack that you are popping from..
Can it pop until it finds a NULL?

Reply:
> Can it pop until it finds a NULL?

Not generally speaking.  NULL is a perfectly valid pointer to push, see...  we know that this isn't the case here, but that knowledge could be fragile, so I'd rather not trust that.

Reply:
But this isn't general, this is a specific stack for the (de)serializer.  What does other parts of the code do if it pops a NULL?  Better, simpler, easier to make the invariant be NULL not allowed.